__tests__/services/TabEvents.spec.ts

```diff
@@ -0,0 +1,394 @@
+      expect(spyTabEvents.cleanFromTabEvents).toHaveBeenCalledTimes(1);
+    });
+
+    it('should sanitize favIconUrl if debug was enabled', () => {
+      TestStore.changeSetting('discardedCleanup', true);
+      TestStore.changeSetting('debugMode', true);
+      TabEvents.onTabDiscarded(0, {...sampleChangeInfo}, sampleTab);
+      expect(spyLib.cadLog.mock.calls[0][0].x.changeInfo.favIconUrl).toBe('***');
+      expect(spyTabEvents.cleanFromTabEvents).not.toHaveBeenCalled();
+    });
+  });
+
+  describe('onTabUpdate', () => {
+    beforeAll(() => {
+      when(spyTabEvents.getAllCookieActions)
+        .calledWith()
+        .mockResolvedValue(undefined as any);
+    });
+    afterAll(() => {
+      spyTabEvents.getAllCookieActions.mockRestore();
+    });
+
+    it('should do nothing if tab status is not "complete"', () => {
+      TabEvents.onTabUpdate(0, sampleChangeInfo, {...sampleTab, status: 'loading'});
+      expect(spyTabEvents.getAllCookieActions).not.toHaveBeenCalled();
+    });
+
+    it('should trigger getAllCookieActions', () => {
+      TabEvents.onTabUpdate(0, sampleChangeInfo, { ...sampleTab, status: 'complete' });
+      jest.runAllTimers();
+      expect(spyTabEvents.getAllCookieActions).toHaveBeenCalledTimes(1);
+    });
+
+    it('should sanitize favIconUrl if status=complete and debug is true', () => {
+      TestStore.changeSetting('debugMode', true);
+      TabEvents.onTabUpdate(0, sampleChangeInfo, { ...sampleTab, status: 'complete' });
+      jest.runAllTimers();
+      expect(spyLib.cadLog.mock.calls[0][0].x.changeInfo.favIconUrl).toBe('***');
+    });
+
+    it('should not queue getAllCookieActions if one is pending already', () => {
+      TestStore.changeSetting('debugMode', true);
+      TabEvents.onTabUpdate(0, sampleChangeInfo, { ...sampleTab, status: 'complete' });
+      expect(spyLib.cadLog.mock.calls[0][0].msg.indexOf('set')).not.toBe(-1);
+      TabEvents.onTabUpdate(0, sampleChangeInfo, { ...sampleTab, status: 'complete' });
+      expect(spyLib.cadLog.mock.calls[1][0].msg.indexOf('pending')).not.toBe(-1);
+      jest.runAllTimers();
+    });
+  });
+
+  describe('onDomainChange', () => {
+    // Because it is not easy checking private variables, use cadLog's inbound msg
+    // to check, even though it is not being output if debug=false.
+    //       spyLib.cadLog.mock.calls[0][0] for first call & first argument
+
+    // Do not change any of the test order as each test relies on the previous actions.
+
+    it('should do nothing if tab.status is not complete', () => {
+      TabEvents.onDomainChange(0, sampleChangeInfo, {...sampleTab, status: 'loading'});
+      expect(spyTabEvents.cleanFromTabEvents).not.toHaveBeenCalled();
+    });
+
+    it('should set mainDomain on first encounter', () => {
+      TabEvents.onDomainChange(0, sampleChangeInfo, {...sampleTab, status: 'complete'});
+      expect(spyLib.cadLog.mock.calls[0][0].msg.indexOf('First')).not.toBe(-1);
```

Comment:
Like before, I'm not sure if we need spy libs for cadLogs and test for whether or not it was called, since it would mean the TabEvents class would depend on on cadLog.

Wouldn't it be the same coverage since cadLog would be unit tested independently?

Same comments whenever spylib for cadLog gets called.

Reply:
For onDomainChange at least should be needed as there is no other way of expecting something as the tabToDomain variable is set to private and jest can't check private variables unless I were to make another temporary class that extends TabEvents just to check/expose that private variable...

While I can just call the functions for coverage, there's not really a point unless something is expected from it.

Reply:
Also I'm not just checking if cadLog was called but the parameters passed into it.

Reply:
I'll try the temporary class method just like what I did for `StoreUser` in the morning.

Reply:
Mostly resolved through 6f01422.  

There are still (3) calls to spyLib.cadLog due to the fact that I'm sanitizing a variable that is useless in debug output.